cmd/lakectl/cmd/ingest.go

```diff
@@ -21,51 +49,77 @@ var ingestCmd = &cobra.Command{
 		dryRun := MustBool(cmd.Flags().GetBool("dry-run"))
 		from := MustString(cmd.Flags().GetString("from"))
 		to := MustString(cmd.Flags().GetString("to"))
+		concurrency := MustInt(cmd.Flags().GetInt("concurrency"))
 		lakefsURI := MustParsePathURI("to", to)
 
+		// initialize worker pool
+		var wg sync.WaitGroup
+		wg.Add(concurrency)
+		requests := make(chan *stageRequest)
+		responses := make(chan *stageResponse)
+		for w := 0; w < concurrency; w++ {
+			go stageWorker(&wg, requests, responses)
+		}
+
 		summary := struct {
 			Objects int64
 			Bytes   int64
 		}{}
-		client := getClient()
-		err := store.Walk(ctx, from, func(e store.ObjectStoreEntry) error {
-			if dryRun {
-				Fmt("%s\n", e)
+
+		// iterate entries and feed our pool
+		go func() {
+			err := store.Walk(ctx, from, func(e store.ObjectStoreEntry) error {
+				if dryRun {
+					Fmt("%s\n", e)
+					return nil
+				}
+				key := e.RelativeKey
+				if lakefsURI.Path != nil && *lakefsURI.Path != "" {
+					path := *lakefsURI.Path
+					if strings.HasSuffix(*lakefsURI.Path, "/") {
+						key = path + key
+					} else {
+						key = path + "/" + key
+					}
+				}
+				mtime := e.Mtime.Unix()
+				requests <- &stageRequest{
+					ctx:        ctx,
+					repository: lakefsURI.Repository,
+					branch:     lakefsURI.Ref,
+					params: &api.StageObjectParams{
+						Path: key,
+					},
+					body: api.StageObjectJSONRequestBody{
+						Checksum:        e.ETag,
+						Mtime:           &mtime,
+						PhysicalAddress: e.Address,
+						SizeBytes:       e.Size,
+					},
+				}
 				return nil
+			})
+			if err != nil {
+				DieFmt("error walking object store: %v", err)
 			}
-			key := e.RelativeKey
-			if lakefsURI.Path != nil && *lakefsURI.Path != "" {
-				path := *lakefsURI.Path
-				if strings.HasSuffix(*lakefsURI.Path, "/") {
-					key = path + key
-				} else {
-					key = path + "/" + key
-				}
+			close(requests)  // we're done feeding work!
+			wg.Wait()        // until all responses have been written
+			close(responses) // so we're also done with responses
+		}()
+
+		for response := range responses {
+			DieOnResponseError(response.resp, response.error)
+			summary.Objects += 1
+			summary.Bytes += api.Int64Value(response.resp.JSON201.SizeBytes)
+
+			// update every 10k records
+			if summary.Objects%10000 == 0 {
+				Write("Staged {{ .Objects | green }} objects so far...\n", summary)
```

Comment:
```suggestion
				Write("Staged {{ .Objects | green }} objects so far...\r", summary)
```
might give a cute single-line "animation".  (But you'd have to write that `\n` later, when done)

Reply:
Nice idea. Done.